### ISSUE TITLE
e2e test framework expose options for pod PriorityClass and quota

### ISSUE THREAD

**Issue by rackeric** [kubernetes/kubernetes#114006]
### What would you like to be added?

- expose option for custom PriorityClass used in test Pods
- expose option to add custom quota to test namespaces

### Why is this needed?

Some clusters require a PriorityClass be set on all Pods, and if setting a PriorityClass then you may need to set a quota in test namespaces.

If there are other ways around this I'm interested in hearing how others get around this.

We have this enabled in clusters: [Limit Priority Class consumption by default](https://kubernetes.io/docs/concepts/policy/resource-quotas/#limit-priority-class-consumption-by-default)

**Comment by pohly** [comment]
/cc

As [discussed on Slack](https://kubernetes.slack.com/archives/C09QZ4DQB/p1668796862379909), the [`KUBE_TEST_REPO_LIST`](https://github.com/kubernetes/kubernetes/blob/57eb5d631ccd615cd161b6da36afc759af004b93/test/utils/image/manifest.go#L74) env variable should allow reconfiguration of all images.

The other two points are valid.

**Comment by aojea** [comment]
> * expose option for custom PriorityClass used in test Pods
> * expose option to add custom quota to test namespaces

I don't understand these 2 options in the context of e2e tests, since this is influencing the environment of the tests, but I may be missing something

e2e tests are meant to test kubernete, some of them are also used to guarantee conformance, but all the tests are meant to be used in kubernetes  environments without any additional enforcements, per example, you can not have firewalls , ...

**Comment by pohly** [comment]
> this is influencing the environment of the tests

I see it the other way around: the environment of the tests (= the cluster) has an influence on what is possible in the tests (= pods must have PriorityClass). It is debatable whether a cluster which requires that field to be set can pass conformance tests, because a conformant cluster should be able to run pods without it - probably not.

But when ignoring that, it makes sense to me to reuse the tests to exercise, say, storage functionality even when that only works by creating pods a bit differently than the test suite normally would do.

I don't know what the custom quota is meant to achieve, but as the e2e framework creates the namespaces, that's where that would have to be configurable.

**Comment by aojea** [comment]
I don't think that we should convert the e2e tests in a customizable framework for kubernetes testing,  the e2e should test kubernetes ... we have the https://github.com/kubernetes-sigs/e2e-framework for those other things

**Comment by aojea** [comment]
coming back to the request, and following up on the slack discussion here
https://kubernetes.slack.com/archives/C09QZ4DQB/p1668798086510869?thread_ts=1668796862.379909&cid=C09QZ4DQB

As an example, when pod security admission went GA the possibility to add labels to the namespaces was added to set the security level, otherwise tests will start to fail on CI

https://github.com/kubernetes/kubernetes/pull/106454

> expose option for custom PriorityClass used in test Pods

I don't see how you can achieve this, since every test create their own pods, and some of the test may be influenced by setting this option, there is not an unified way to create pods

> expose option to add custom quota to test namespaces

e2e test may run in parallel or in serial, having an option that create quotas when running in parallel may add contention and make some tests to fail or timeout

**Comment by pohly** [comment]
> the e2e should test kubernetes ... 

I think that is what @rackeric is trying to do.

Let's not confuse test/e2e/framework with test/e2e:
- test/e2e/framework: internal to Kubernetes, not meant to be used outside
- test/e2e: part of the release artifacts, to be used outside of Kubernetes, in particular conformance testing

This issue is about how to use test/e2e. That test/e2e/framework and tests need to be modified is an implementation detail.

> we have the https://github.com/kubernetes-sigs/e2e-framework for those other things

That doesn't have tests and therefore is no replacement for test/e2e.

**Comment by aojea** [comment]
My understanding is different but is it worth to clarify,

> * test/e2e: part of the release artifacts, to be used outside of Kubernetes, in particular conformance testing

AFAIk,only conformance tests are supposed to be consumed outside, and they are only a subsets of the e2e tests https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md

`test/e2e` is not part of the release and is internal and only used for developer and kubernetes testing https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/e2e-tests.md

But I may be wrong, is better if @spiffxp and @BenTheElder can chime in

**Comment by pohly** [comment]
> AFAIk,only conformance tests are supposed to be consumed outside, 

SIG Storage (okay, me with a different hat on, but it has been fully embraced by SIG Storage and Red Hat for their CSI driver testing on OpenShift) also uses the in-tree storage tests as the recommended set of tests that out-of-tree CSI drivers should pass to validate their behavior.

> test/e2e is not part of the release 

It is. The e2e.test binary built from it is included in release tar balls.

**Comment by aojea** [comment]
> It is. The e2e.test binary built from it is included in release tar balls.

I can't see in the release notes

https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.25.md#downloads-for-v1254

it was, wasn't it?

https://github.com/kubernetes/enhancements/tree/master/keps/sig-testing/714-break-test-tarball

**Comment by rackeric** [comment]
I updated the title of this to exclude the docker repo ask... So I found out that a default PriorityClass is being set when PodSpec does not specify this, so this is no longer a requirement but should be a nice to have since it could help make sure the pods get scheduled by setting higher priority.

However, what really is needed is a ResourceQuota in the test namespace stating that you can use the selected Pod PriorityClass, even if just using the default set by the cluster.

So, the current main question is how the community recommend running e2e tests when the cluster has a ResourceQuota admission plugin enabled? Disable it?/Pre-provisioning?

**Comment by aojea** [comment]
I can't see how you can reliable run e2e tests with the ResourceQuota admission plugin enabled, tests are added continuously by different SIGs and I didn't see any place where we track the number of resources requested , how do you estimate those quotas in advance? ... maybe someone is running the tests with the plugin enabled , that feedback will be really nice

**Comment by liggitt** [comment]
> I can't see how you can reliable run e2e tests with the ResourceQuota admission plugin enabled

To clarify, just enabling the resourcequota admission plugin does not prevent conformance tests from running (it's enabled in all clusters we run CI tests on in Kubernetes).

Restricting quota by default with a cluster-wide configuration is what is causing difficulties for the tests. There was some additional discussion in the [slack thread](https://kubernetes.slack.com/archives/C09QZ4DQB/p1668796862379909) about possible ways to run conformance tests on a cluster like that.

**Comment by pohly** [comment]
So from a conformance perspective, the e2e tests are not *required* to pass in the cluster set up by @rackeric. But I can see that it would still be *useful* and if it is not too burdensome for us as upstream maintainers, then I think we should accept patches that enable additional usages of the e2e test suite. This will be a case-by-case decision.

@rackeric: do you want to get changes into the upstream e2e suite or is patching it locally sufficient for you?

**Comment by liggitt** [comment]
> I think we should accept patches that enable additional usages of the e2e test suite

that can be reasonable, but we want to be careful in a couple dimensions:
* rather than adding lots of specific custom aspects into the e2e to allow interoperability with custom external config, a general solution like "wait for labels x,y,z on newly created namespaces before proceeding" would allow external namespace controllers to interact with the e2e runs without baking a lot of logic into e2e binaries
* the conformance tests especially are intended to demonstrate what behavior end users can expect from a system… making end users to set up resource quotas for themselves in every namespace isn't something I would expect, so having the e2e test do that doesn't seem great

**Comment by aojea** [comment]
> To clarify, just enabling the resourcequota admission plugin does not prevent conformance tests from running (it's enabled in all clusters we run CI tests on in Kubernetes).

thanks for clarifying, I was with a wrong assumption then 😅 

> wait for labels x,y,z on newly created namespaces before proceeding

There are some tests that create multiple namespaces, that will not work for those or will make the logic complicated to support them.

**Comment by liggitt** [comment]
> There are some tests that create multiple namespaces, that will not work for those or will make the logic complicated to support them.

If all namespace creations funnel through the e2e namespace creation function (they should), it's easy for that function to wait for a label on the namespace before returning

**Comment by aojea** [comment]
you are right, it seems it was fixed 3 days ago https://github.com/kubernetes/kubernetes/issues/108298

**Comment by aojea** [comment]
> * rather than adding lots of specific custom aspects into the e2e to allow interoperability with custom external config, a general solution like "wait for labels x,y,z on newly created namespaces before proceeding" would allow external namespace controllers to interact with the e2e runs without baking a lot of logic into e2e binaries

+1

**Comment by aojea** [comment]
Let's the naming bikeshedding start, how do we want to tag the e2e test namespaces ...

`e2e.test.k8s.io: sig-...`

**Comment by liggitt** [comment]
oh, that wasn't what I meant... I meant to allow the e2e test to be invoked with an option to wait for label(s)/annotation(s) to show up on namespaces it creates before considering those namespaces "ready", and for the external cluster-provided controller to do the namespace annotating/labeling

**Comment by aojea** [comment]
1. e2e starts running
2. e2e runs test 1 and creates namespace foo for that test
3. e2e waits until namespace foo has annotation/label X ?
4. e2e starts testing or timeout
??

**Comment by liggitt** [comment]
Basically, yeah. Something sort of like this in the CreateTestingNS function:

```diff
diff --git a/test/e2e/framework/util.go b/test/e2e/framework/util.go
index 964dbc6fc2f..d2abfaffb53 100644
--- a/test/e2e/framework/util.go
+++ b/test/e2e/framework/util.go
@@ -377,6 +377,13 @@ func CreateTestingNS(ctx context.Context, baseName string, c clientset.Interface
 			return got, err
 		}
 	}
+
+	if TestContext.ReadyNamespaceSelector != nil {
+		if err := WaitForNamespaceSelector(ctx, TestContext.ReadyNamespaceSelector, got.Name); err != nil {
+			return got, err
+		}
+	}
+
 	return got, nil
 }
```

**Comment by aojea** [comment]
I naively thought that ResourceQuota and external components could identify `namespaces` used by the `e2e` tests and act based on that, as "I can not restrict anything here, let this client do his thing"

I'm afraid this behaviour may impact some tests, at least the ones that create multiple namespaces  ... also, do we want to support that behaviour, it looks like opening a whole can of worms ...

**Comment by enj** [comment]
> I naively thought that ResourceQuota and external components could identify `namespaces` used by the `e2e` tests and act based on that, as "I should not restrict anything here, let this client do his thing"

Sounds like a great way for a hacker to bypass quota restrictions :p

**Comment by aojea** [comment]
I assume you configure your cluster to only trust e2e tests when you are running e2e tests ;)
